### PR TITLE
[WIP]Orca: Add ncf_orca.py with spark dataframe

### PR DESCRIPTION
**Pull request by Chaselzxy**
## Description

Read the data set in `ncf_orca.py` into spark dataframe, process it, generate negative samples, and change the model to a binary model.

### 1. Why the change?
Support the operation of data in spark dataframe.

### 3. Summary of the change 

1. Read the data in the data set(.csv) into spark dataframe.
2. Perform data processing and statistics on the data in spark dataframe.
3. Generate several negative samples.
4. Change the model to a binary model.

### 4. How to test?
- [ ] Document test

### CODE REVIEW

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import time
+import numpy as np
+import tensorflow as tf
+import os
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:20]: Pls check import order.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+from bigdl.orca import OrcaContext, init_orca_context, stop_orca_context
+from bigdl.orca.learn.tf2 import Estimator
+
+
+def model_creator(config):
+    import tensorflow as tf
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:27]: redundant import

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    
+    model = keras.Model(inputs=[user, item], outputs=outputs)
+    model.compile(optimizer= "adam",
+                  loss= "binary_crossentropy",
+                  metrics=['accuracy'])
+    return model
+
+     
+
+cluster_mode = "local"
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:63]: Maybe add command-line options, arguments.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+                           container_image="10.239.45.10/arda/intelanalytics/bigdl-k8s-spark-3.1.2:0.14.0-SNAPSHOT",
+                           conf={"spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data"})
+elif cluster_mode == "spark-submit":  # To test k8s using spark-submit
+    sc = init_orca_context(cluster_mode="spark-submit")
+
+
+# data_path = "/home/kai/Downloads"
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:83]: Pls remove unnecessary comments.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    return model
+
+     
+
+cluster_mode = "local"
+# cluster_mode = "k8s"
+if cluster_mode == "local":
+    sc = init_orca_context(memory="20g")
+elif cluster_mode == "standalone":
+    sc = init_orca_context("standalone", master="spark://...", cores=8, num_nodes=4, memory="10g")
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:68]: Do not need parameter `master`, it is needed in k8s related cluster mode.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+elif cluster_mode == "yarn":
+    sc = init_orca_context("yarn", cores=8, num_nodes=4, memory="10g")
+elif cluster_mode == "k8s":
+    sc = init_orca_context(cluster_mode="k8s", cores=8, num_nodes=4, memory="20g",
+                           master="k8s://https://172.16.0.200:6443",
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:73]: As a tutorial, I think it is better to put the hard-coded parameters in command line options and arguments.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+# backend = 'spark'
+# est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1, backend=backend, model_dir=model_dir)
+est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1)
+
+stats = est.fit(train_df,
+                epochs=epochs,
+                batch_size=batch_size,
+                feature_cols=['user', 'item'],
+                label_cols=['label'],
+                steps_per_epoch=int(800000 // batch_size),
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:145]: Maybe use variable to replace numbers to enhance code readability.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+# evaluate with Estimator
+stats = est.evaluate(test_df, 
+                     feature_cols=['user', 'item'],
+                     label_cols=['label'],
+                     num_steps=int(0.2*num_sample // batch_size))
+
+print(stats)
+est.shutdown()
+end = time.time()
+print("Time used: ", end - start)
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:167]: May I ask what time information is counted here?

**Comment by lalalapotter** [comment]
Do we need a README.md for each tutorial scripts?

**Comment by hkvision** [comment]
> Do we need a README.md for each tutorial scripts?

Not at this moment. Probably add a guide after all tutorials are finished :)

**Review by lalalapotter — CHANGES_REQUESTED** [review]

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+#
+
+import sys
+import time
+import os
+
+import tensorflow as tf
+from tensorflow import keras
+
+from bigdl.orca import OrcaContext, init_orca_context, stop_orca_context
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:24]: put OrcaContext after stop_orca_context

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+from bigdl.orca.learn.tf2 import Estimator
+
+
+def model_creator(config):
+    embedding_size=16
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:29]: put embedding size into config?

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    user = keras.layers.Input(dtype=tf.int32, shape=(None,))
+    item = keras.layers.Input(dtype=tf.int32, shape=(None,))
+    label = keras.layers.Input(dtype=tf.int32, shape=(None,))
+
+    with tf.name_scope("GMF"):
+        user_embed_GMF = keras.layers.Embedding(max_user_id + 1, embedding_size)(user)
+        item_embed_GMF = keras.layers.Embedding(max_item_id + 1, embedding_size)(item)
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:36]: put max_user/item id into config?

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 6 above)
Comment (hkvision) [reply python/orca/tutorial/tensorflow/ncf_orca.py:71]: right, use argparser

Comment (hkvision) [reply python/orca/tutorial/tensorflow/ncf_orca.py:71]: do not hardcode k8s master

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+data_type = "ml-1m"
+# Need spark3 to support delimiter with more than one character.
+start = time.time()
+spark= OrcaContext.get_spark_session()
+from pyspark.sql.types import StructField, StructType, IntegerType
+
+schema = StructType(
+    [
+        StructField('_c0', IntegerType(), True),
+        StructField('_c1', IntegerType(), True)
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:91]: You can just name the columns to user and item here and remove the rename in Line 96

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 8 above)
Comment (hkvision) [reply python/orca/tutorial/tensorflow/ncf_orca.py:158]: I think we can just remove the timing in the tutorial?

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+    ]
+)
+df0 = spark.read.csv("{}/{}/ratings.csv".format(data_path, data_type), sep="::",schema=schema,
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:94]: just use df not df0

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1)
+
+stats = est.fit(train_df,
+                epochs=epochs,
+                batch_size=batch_size,
+                feature_cols=['user', 'item'],
+                label_cols=['label'],
+                steps_per_epoch=int(0.8*num_sample // batch_size),
+                validation_data=test_df,
+                validation_steps =int(0.2*num_sample // batch_size))
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:143]: should use train_df/test_df.count() instead of calculating by yourself

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+checkpoint_path = os.path.join(model_dir, "NCF.ckpt")
+est.save_checkpoint(checkpoint_path)
+
+
+# evaluate with Estimator
+stats = est.evaluate(test_df,
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:150]: add batch_size

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+print(min_user_id, max_user_id, min_item_id, max_item_id)
+from pyspark.sql import functions
+df = df0.withColumn('label', functions.lit(1))
+
+from pyspark.sql import functions as F
+from pyspark.sql import types as T
+import random
+
+def neg_sample(x):
+    item_count = len(x) * 4
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:111]: The number of negative samples for each (user, item) pair should be modifiable, and you can specify it in arguments as well.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    max_count = max_item_id - len(set(x))
```
Comment (lalalapotter) [inline python/orca/tutorial/tensorflow/ncf_orca.py:112]: Could you please explain the upper bound here?

Comment (Chaselzxy) [reply python/orca/tutorial/tensorflow/ncf_orca.py:109]: Some users give scores to a large number of items, resulting in the number of remaining items being less than four times the number already scored.

Comment (Chaselzxy) [reply python/orca/tutorial/tensorflow/ncf_orca.py:109]: > Could you please explain the upper bound here?

Some users give scores to a large number of items, resulting in the number of remaining items being less than four times the number already scored.

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 8 above)
Comment (lalalapotter) [reply python/orca/tutorial/tensorflow/ncf_orca.py]: Sure.

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 17 above)
Comment (lalalapotter) [reply python/orca/tutorial/tensorflow/ncf_orca.py:109]: Whether there are duplicate samples in negative samples?

Comment (Chaselzxy) [reply python/orca/tutorial/tensorflow/ncf_orca.py:109]: not at the present

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+min_user_id = df.agg({"user": "min"}).collect()[0]["min(user)"]
+max_user_id = df.agg({"user": "max"}).collect()[0]["max(user)"]
+min_item_id = df.agg({"item": "min"}).collect()[0]["min(item)"]
+max_item_id = df.agg({"item": "max"}).collect()[0]["max(item)"]
+print(min_user_id, max_user_id, min_item_id, max_item_id)
+from pyspark.sql import functions
+df = df.withColumn('label', functions.lit(1))
+
+from pyspark.sql import functions as F
+from pyspark.sql import types as T
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:103]: from pyspark.sql.functions import udf, ...

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+spark= OrcaContext.get_spark_session()
+from pyspark.sql.types import StructField, StructType, IntegerType
+
+schema = StructType(
+    [
+        StructField('user', IntegerType(), True),
+        StructField('item', IntegerType(), True)
+    ]
+)
+df = spark.read.csv("{}/{}/ratings.csv".format(data_path, data_type), sep="::",schema=schema,
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:92]: why using csv not dat?

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+    item_count = len(x) * neg_scale
+    max_count = max_item_id - len(set(x))
+    neg_item = random.sample(set(range(min_item_id, max_item_id+1)) - set(x), min(item_count,max_count))
+    return neg_item
+
+neg_sample_udf = F.udf(neg_sample, T.ArrayType(IntegerType(), False))
+
+df_neg= df.groupBy('user').agg(neg_sample_udf(F.collect_list('item')).alias('item_list'))
+from pyspark.sql.functions import *
+df_neg = df_neg.select(df_neg.user, explode(df_neg.item_list))
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:117]: will the column have name item?

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+df_neg = df_neg.withColumn('label', functions.lit(0))
+df = df.unionAll(df_neg)
+num_sample = df.count()
+train_df, test_df = df.randomSplit([0.8, 0.2],100)
+
+
+batch_size=256
+epochs=5
+model_dir='./'
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:126]: is it used?

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1)
+
+stats = est.fit(train_df,
+                epochs=epochs,
+                batch_size=batch_size,
+                feature_cols=['user', 'item'],
+                label_cols=['label'],
+                steps_per_epoch=int(60000 // batch_size),
+                validation_data=test_df,
+                validation_steps =int(10000 // batch_size))
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:140]: should not hard code 60000 and 10000, use df.count

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+
+# save model in H5 format
+est.save("/tmp/cifar10_model.h5", save_format='h5')
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:143]: model name is wrong, also don't save to temp, just save to the current dir

Comment (hkvision) [reply python/orca/tutorial/tensorflow/ncf_orca.py:143]: I think we don't need to add save_format using the latest BigDL?

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 17 above)
Comment (hkvision) [reply python/orca/tutorial/tensorflow/ncf_orca.py:109]: It makes sense to do negative sampling without duplication.

File: python/orca/tutorial/tensorflow/ncf_orca.py [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+                           container_image="10.239.45.10/arda/intelanalytics/bigdl-k8s-spark-3.1.2:0.14.0-SNAPSHOT",
+                           conf={"spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data"})
+elif cluster_mode == "spark-submit":  # To test k8s using spark-submit
+    sc = init_orca_context(cluster_mode="spark-submit")
+
+
+data_path = "./bigdl2.0/data"
```
Comment (hkvision) [inline python/orca/tutorial/tensorflow/ncf_orca.py:80]: change this

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 20 above)
Comment (Chaselzxy) [reply python/orca/tutorial/tensorflow/ncf_orca.py:117]: > 

it won't have name, but its name is not important, because it will be unionAll after df

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 22 above)
Comment (Chaselzxy) [reply python/orca/tutorial/tensorflow/ncf_orca.py:125]: > should not hard code 60000 and 10000, use df.count

When step is set to 60000/10000 (refer to tutorial), the validation accuracy is ideal. When df.count is used, the result is abnormal.

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 24 above)
Comment (Chaselzxy) [reply python/orca/tutorial/tensorflow/ncf_orca.py:66]: > change this

What path should we read data from?

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 22 above)
Comment (hkvision) [reply python/orca/tutorial/tensorflow/ncf_orca.py:125]: Then you need to figure out why it is abnormal instead of hard code the values.

File: python/orca/tutorial/tensorflow/ncf_orca.py (hunk 24 above)
Comment (hkvision) [reply python/orca/tutorial/tensorflow/ncf_orca.py:66]: Just the currently directory? To align with the PyTorch example?
